ipie/addons/ephqmc/trial_wavefunction/variational/toyozawa_variational.py

```diff
@@ -0,0 +1,166 @@
+import numpy as np
+from scipy.optimize import minimize, basinhopping 
+from ipie.systems import Generic
+from ipie.addons.eph.hamiltonians.holstein import HolsteinModel
+
+from jax.config import config
+config.update("jax_enable_x64", True)
+import jax
+import jax.numpy as npj
+
+def gab(A,B):
+    inv_O = npj.linalg.inv((A.conj().T).dot(B))
+    GAB = B.dot(inv_O.dot(A.conj().T))
+    return GAB
+
+def gradient_toyozawa_mo(x, nbasis, nup, ndown, T, U, g, m, w0, perms, restricted):
+    grad = np.array(jax.grad(objective_function_toyozawa_mo)(x, nbasis, nup, ndown, T, U, g, m, w0, perms, restricted))
+    return grad
+
+def objective_function_toyozawa_mo(x, nbasis, nup, ndown, T, U, g, m, w0, perms, restricted, zero_th=1e-12):
+    nbasis = int(round(nbasis))
+    nup = int(round(nup))
+    ndown = int(round(ndown))
+    
+    shift0 = x[:nbasis]
+
+    nbsf = nbasis
+    nocca = nup
+    noccb = ndown
+    nvira = nbasis - nocca
+    nvirb = nbasis - noccb
+    
+    psi0a = npj.array(x[nbsf:nbsf+nbsf*nocca],dtype=npj.float64)
+    psi0a = psi0a.reshape((nocca, nbsf)).T
+
+    if (noccb>0):
+        psi0b = npj.array(x[nbsf+nbsf*nocca : ], dtype=npj.float64)
+        psi0b = psi0b.reshape((noccb, nbsf)).T
+    
+    nperms = len(perms)
+
+    num_energy = 0.
+    denom = 0.0
+        
+    beta0 = shift0 * npj.sqrt(m * w0 /2.0)
+
+    for i,perm in enumerate(perms):    
+        psia_j = psi0a[perm, :]
+        beta_j = beta0[npj.array(perm)]
+        
+        if (noccb > 0):
+            psib_j = psi0b[perm, :]
+            overlap = npj.linalg.det(psi0a.T.dot(psia_j)) * npj.linalg.det(psi0b.T.dot(psib_j)) * npj.prod(npj.exp(-0.5 * (beta0**2 + beta_j**2) + beta0*beta_j))
+        else:
+            overlap = npj.linalg.det(psi0a.T.dot(psia_j)) * npj.prod(npj.exp (- 0.5 * (beta0**2 + beta_j**2) + beta0*beta_j))
+
+        if overlap < zero_th:
+            continue
+
+        Ga_j = gab(psi0a, psia_j)
+        if (noccb > 0):
+            Gb_j = gab(psi0b, psib_j)
+        else:
+            Gb_j = npj.zeros_like(Ga_j)
+
+        G_j = [Ga_j, Gb_j]
+
+        rho = G_j[0].diagonal() + G_j[1].diagonal()
+        ke = npj.sum(T[0] * G_j[0] + T[1] * G_j[1])
+        pe = U * npj.dot(G_j[0].diagonal(), G_j[1].diagonal())
+        e_ph = w0 * npj.sum(beta0 * beta_j) 
+        e_eph = - g * npj.dot(rho, beta0 + beta_j)
+
+        num_energy += (ke + pe + e_ph + e_eph) * overlap  # 2.0 comes from hermiticity
+        denom += overlap
+
+    etot = num_energy / denom
+    return etot.real
+
+def circ_perm(lst):
```

Comment:
docstring?  What is this doing, is there no numpy equivalent?

Reply:
I could write something equivalent with vstack and roll, which may improve readability I suppose?

Reply:
Hmm, just write a docstring and a unit test which demonstrates what the function is supposed to do given some set of inputs so that someone in the future can figure things out.